### ISSUE TITLE
[release-1.20] Empty label pods stuck in ContainerCreating

### ISSUE THREAD

**Issue by ClaudiuLSH** [rancher/rke2#1530]
**Environmental Info:**
RKE2 Version: `v1.20.9+rke2r1`

Node(s) CPU architecture, OS, and Version: `Linux k8s18-stg 5.8.0-63-generic #71~20.04.1-Ubuntu SMP Thu Jul 15 17:46:08 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux`

Cluster Configuration: 3 servers

**Describe the bug:**
Pods with empty labels will be stuck in `ContainerCreating`. This might be kubernetes version related but I am not sure at this point.

**Steps To Reproduce:**
1. Install rke2 v1.20.9+rke2r1 (this bug is not present on v1.20.8+rke2r1);
2. Deploy a pod with empty label.
Example deployment manifest:
```
apiVersion: apps/v1
kind: Deployment
metadata:
  name: nginx-deployment
  labels:
    app: nginx
spec:
  replicas: 10
  selector:
    matchLabels:
      app: nginx
  template:
    metadata:
      labels:
        app: nginx
        test: ""
    spec:
      containers:
      - name: nginx
        image: nginx:1.14.2
        ports:
        - containerPort: 80
```

**Additional context / logs:**
/var/lib/rancher/rke2/agent/containerd/containerd.log
```
time="2021-08-04T09:40:10.638690941Z" level=error msg="RunPodSandbox for &PodSandboxMetadata{Name:nginx-deployment-688974dccf-5p5xm,Uid:d3dc35b1-448a-46e8-a684-1d30087fd26b,Namespace:default,Attempt:0,} failed, error" error="failed to setup network for sandbox \"ac36e4080e51ace591daced877f164a744e276742cd88c0ae176e062a37f3e0f\": Unable to create endpoint: response status code does not match any response statuses defined for this endpoint in the swagger spec (status 429): {}"
level=warning msg="Errors encountered while deleting endpoint" error="[DELETE /endpoint/{id}][404] deleteEndpointIdNotFound " subsys=cilium-cni
time="2021-08-04T09:40:11.632437037Z" level=error msg="RunPodSandbox for &PodSandboxMetadata{Name:nginx-deployment-688974dccf-thxzh,Uid:cb29ce38-18be-4146-a1ac-0b699afd207b,Namespace:default,Attempt:0,} failed, error" error="failed to setup network for sandbox \"86b08a1723ced573379f37f2b105e4c6282b5ad8b112d9e8276d00e9cad5e242\": Unable to create endpoint: response status code does not match any response statuses defined for this endpoint in the swagger spec (status 429): {}"
time="2021-08-04T09:40:21.477621543Z" level=info msg="RunPodsandbox for &PodSandboxMetadata{Name:nginx-deployment-688974dccf-5p5xm,Uid:d3dc35b1-448a-46e8-a684-1d30087fd26b,Namespace:default,Attempt:0,}"
time="2021-08-04T09:40:24.476971661Z" level=info msg="RunPodsandbox for &PodSandboxMetadata{Name:nginx-deployment-688974dccf-thxzh,Uid:cb29ce38-18be-4146-a1ac-0b699afd207b,Namespace:default,Attempt:0,}"
level=warning msg="Errors encountered while deleting endpoint" error="[DELETE /endpoint/{id}][404] deleteEndpointIdNotFound " subsys=cilium-cni
time="2021-08-04T09:40:36.669911388Z" level=error msg="RunPodSandbox for &PodSandboxMetadata{Name:nginx-deployment-688974dccf-5p5xm,Uid:d3dc35b1-448a-46e8-a684-1d30087fd26b,Namespace:default,Attempt:0,} failed, error" error="failed to setup network for sandbox \"5ea14771a91d51351516a8260d94c8dd2e1e2fddec8cbf1e3e02e5a2a9c01e21\": Unable to create endpoint: response status code does not match any response statuses defined for this endpoint in the swagger spec (status 429): {}"
level=warning msg="Errors encountered while deleting endpoint" error="[DELETE /endpoint/{id}][404] deleteEndpointIdNotFound " subsys=cilium-cni
time="2021-08-04T09:40:39.656021271Z" level=error msg="RunPodSandbox for &PodSandboxMetadata{Name:nginx-deployment-688974dccf-thxzh,Uid:cb29ce38-18be-4146-a1ac-0b699afd207b,Namespace:default,Attempt:0,} failed, error" error="failed to setup network for sandbox \"ab675fd69777b6d8436b4a9bda058c51249cfd097604e9a5e98d2929b08fb8cb\": Unable to create endpoint: response status code does not match any response statuses defined for this endpoint in the swagger spec (status 429): {}"
level=warning msg="Errors encountered while deleting endpoint" error="[DELETE /endpoint/{id}][404] deleteEndpointIdNotFound " subsys=cilium-cni
```

Cilium logs:
```
level=warning msg="Key allocation attempt failed" attempt=6 error="unable to allocate ID 12439 for key [k8s:app=nginx k8s:io.cilium.k8s.namespace.labels.field.cattle.io/projectId=p-r2sfd k8s:io.cilium.k8s.policy.cluster=default k8s:io.cilium.k8s.policy.serviceaccount=default k8s:io.kubernetes.pod.namespace=default k8s:test]: CiliumIdentity.cilium.io \"12439\" is invalid: security-labels.k8s:test: Invalid value: \"null\": security-labels.k8s:test in body must be of type string: \"null\"" key="[k8s:app=nginx k8s:io.cilium.k8s.namespace.labels.field.cattle.io/projectId=p-r2sfd k8s:io.cilium.k8s.policy.cluster=default k8s:io.cilium.k8s.policy.serviceaccount=default k8s:io.kubernetes.pod.namespace=default k8s:test]" subsys=allocator
```

Anything else I can do to debug this?

**Comment by brandond** [comment]
Containerd is failing because cilium is failing because it doesn't allow empty label values. I would open an issue with Cilium.

**Comment by ClaudiuLSH** [comment]
I would but what is curious is that on v1.20.8+rke2r1 this is working fine. With the same exact cilium version. Do you still think it would be OK to open an issue with cilium?

**Comment by manuelbuil** [comment]
> I would but what is curious is that on v1.20.8+rke2r1 this is working fine. With the same exact cilium version. Do you still think it would be OK to open an issue with cilium?

Interesting... I'll try to have a look tomorrow at this and see if I can reproduce it

**Comment by manuelbuil** [comment]
I was able to reproduce this. I used this manifest:
```
apiVersion: cilium.io/v2
kind: CiliumIdentity
metadata:
  name: "testing"
security-labels:
  k8s-test: "" 
```
when applying it on `v1.20.9`, this is the output
```
The CiliumIdentity "testing" is invalid: security-labels.k8s-test: Invalid value: "null": security-labels.k8s-test in body must be of type string: "null"
```
This does not happen when swapping the kube-api image with the upstream `k8s.gcr.io/kube-apiserver:v1.20.9`. It also does not happen in our v1.20.8, v1.21.3 or master

**Comment by manuelbuil** [comment]
For more information, kube-api expects a `<map[string]string>` :
```
KIND:     CiliumIdentity
VERSION:  cilium.io/v2

FIELD:    security-labels <map[string]string>

DESCRIPTION:
     SecurityLabels is the source-of-truth set of labels for this identity.
```
and from the kubectl logs, I see:
```
Request Body: {"apiVersion":"cilium.io/v2","kind":"CiliumIdentity","metadata":{"annotations":{"kubectl.kubernetes.io/last-applied-configuration":"{\"apiVersion\":\"cilium.io/v2\",\"kind\":\"CiliumIdentity\",\"metadata\":{\"annotations\":{},\"name\":\"testing\"},\"security-labels\":{\"foo\":\"bar\",\"k8s-test\":\"\"}}\n"},"name":"testing"},"security-labels":{"foo":"bar","k8s-test":""}}
```
We are exactly passing that: `"security-labels":{"foo":"bar","k8s-test":""}`

**Comment by brandond** [comment]
This appears to be fixed on kubernetes build with the correct golang version:
```console
[root@centos01 ~]# kubectl version -o yaml
clientVersion:
  buildDate: "2021-08-05T23:06:55Z"
  compiler: gc
  gitCommit: 7a576bc3935a6b555e33346fd73ad77c925e9e4a
  gitTreeState: clean
  gitVersion: v1.20.9+rke2r2
  goVersion: go1.15.14b5
  major: "1"
  minor: "20"
  platform: linux/amd64
serverVersion:
  buildDate: "2021-08-05T23:06:55Z"
  compiler: gc
  gitCommit: 7a576bc3935a6b555e33346fd73ad77c925e9e4a
  gitTreeState: clean
  gitVersion: v1.20.9+rke2r2
  goVersion: go1.15.14b5
  major: "1"
  minor: "20"
  platform: linux/amd64

[root@centos01 ~]# kubectl get ciliumidentity testing -o yaml
apiVersion: cilium.io/v2
kind: CiliumIdentity
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"cilium.io/v2","kind":"CiliumIdentity","metadata":{"annotations":{},"name":"testing"},"security-labels":{"k8s-test":""}}
  creationTimestamp: "2021-08-05T23:57:15Z"
  generation: 1
  name: testing
  resourceVersion: "1018"
  uid: e529c08e-0348-4c29-8763-a1e0e163ba37
security-labels:
  k8s-test: ""
```

**Comment by rancher-max** [comment]
Validated this is fixed in v1.20.9-rc1+rke2r2 following the same steps mentioned in this issue, using cilium as the cni.
